### PR TITLE
Add support for print reftests

### PR DESCRIPTION
**Pull request by jgraham**

### CODE REVIEW

File: infrastructure/reftest/reftest_match-print-ref.html
Comment (gsnedders) [inline infrastructure/reftest/reftest_match-print-ref.html:6]: Where does this 2in come from? But my first thought would be "what's the margin/padding on `@page`?".

Comment (jgraham) [reply infrastructure/reftest/reftest_match-print-ref.html:6]: Page is 5*3 inches with 0.5 inch margins.

Comment (stephenmcgruer) [reply infrastructure/reftest/reftest_match-print-ref.html:6]: Yeah I would try `@page { margin: 0; }` and see if that fixes it.

File: tools/manifest/sourcefile.py
Comment (stephenmcgruer) [inline tools/manifest/sourcefile.py:425]: We didn't spec a directory structure variant in the RFC, did we? Maybe an amendment to make.

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/browsers/chrome.py:86]: Why do we need to put the browser into headless mode for print-reftests?

File: tools/wptrunner/wptrunner/executors/base.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/base.py:83]: So our 'hash' is now a list of hashes; that's probably ok, but are we gaining anything from doing a hexdigest of each screenshot now or should we just return a list of base64 encodings?

Also nit, the docstring is now wrong.

File: tools/wptrunner/wptrunner/executors/base.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/base.py:89]: I haven't reviewed the rest of this file as it seems WIP - let me know if there's any part that needs looked at.

File: tools/wptrunner/wptrunner/executors/executorchrome.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executorchrome.py:29]: I'm not familiar with `\e` here - what is it?

File: tools/wptrunner/wptrunner/executors/executorchrome.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executorchrome.py:43]: I assume then that the chrome-specific endpoint here only accepts inches, not cm? Worth documenting that if so, otherwise the `2.54` is sorta magical.

File: tools/wptrunner/wptrunner/executors/executorchrome.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executorchrome.py:59]: Nit; leftover `print`

File: tools/wptrunner/wptrunner/executors/executorchrome.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executorchrome.py:87]: Nit; https://github.com/web-platform-tests/wpt/issues/7140 is the dpi issue and even then we may want to track dpi support for print-reftests separately from reftests.

File: tools/wptrunner/wptrunner/executors/executormarionette.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executormarionette.py:945]: It seems weird for a method to return a single item in a list, and looking at all the cases where `_screenshot` has been changed that seems to be what is happening. Assuming the caller is gathering a list, could it do the stitching instead?

Also, at least in executorwebdriver.py `_screenshot` wasn't changed, and I'm not sure if that was deliberate or not?

File: tools/wptrunner/wptrunner/executors/executormarionette.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executormarionette.py:982]: Again, same comment re wrapping a single-item list vs whoever is accumulating a list handling it.

File: tools/wptrunner/wptrunner/executors/executormarionette.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executormarionette.py:1114]: Same comment re `dpi` issue

File: tools/wptrunner/wptrunner/executors/executorwebdriver.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executorwebdriver.py:352]: I don't understand what these changes are about; can you explain?

File: tools/wptrunner/wptrunner/wpttest.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/wpttest.py:557]: What is `override` for, if both implementations assert it is `None`?

File: tools/wptrunner/wptrunner/wpttest.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/wpttest.py:601]: So it seems like the webdriver API is in cm, whilst we spec'd this in inches. I presume most paginated media uses inches?

Any thoughts on converting here vs having `viewport_size` in inches and converting close to where we call webdriver?

File: tools/wptserve/wptserve/handlers.py
Comment (stephenmcgruer) [inline tools/wptserve/wptserve/handlers.py:470]: This change may not be wrong, but I'm not clear what it has to do with the rest of the PR?

**Review by stephenmcgruer — COMMENTED** [review]
Looks fairly good at a high level, with some questions. Note that GitHub ate my initial review, so I may have missed some of my original comments when I retyped them all... (GitHubbbbbb! *shakes fist*)

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (jgraham) [reply tools/wptrunner/wptrunner/browsers/chrome.py:86]: Don't know, you tell me ;) But it is apparently a limitation in Chrome that the `printToPDF` CDP endpoint only works in headless.

File: tools/wptrunner/wptrunner/wpttest.py
Comment (jgraham) [reply tools/wptrunner/wptrunner/wpttest.py:601]: I don't think "most paginated media uses inches"; all the A* paper sizes are specified in metric. AFAICT the use of inches by default is basically US-bias in the underlying API. I don't really want to use inches in the external facing parts.

File: tools/wptserve/wptserve/handlers.py
Comment (jgraham) [reply tools/wptserve/wptserve/handlers.py:470]: Without this change we can't setup the print-reftest harness page without passing in empty format args. Note it's a separate commit in the series

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (jgraham) [reply tools/wptrunner/wptrunner/browsers/chrome.py:86]: That sounds like a good question to ask a Blink engineer ;) (aiui it's a limitation of Chrome/CDP that the `printToPDF` endpoint only works in headless mode).

File: tools/wptrunner/wptrunner/executors/executorchrome.py
Comment (jgraham) [reply tools/wptrunner/wptrunner/executors/executorchrome.py:29]: I'm thinking it's a typo that got introduced in 2015 and no one noticed until now.

File: tools/wptrunner/wptrunner/executors/executorwebdriver.py
Comment (jgraham) [reply tools/wptrunner/wptrunner/executors/executorwebdriver.py:352]: It's basically a mechanism to allow the `ChromeDriverPrintReftestExecutor` to inherit from the `WebDriverReftestExector` but also have a special `ChromeDriverProtocol` which is an extension of the `WebDriverProtocol` including the CDP-implemented print parts.

More abstractly it makes the protocol a swappable component rather that an intrinsic part of the executor.

File: tools/wptrunner/wptrunner/wpttest.py
Comment (jgraham) [reply tools/wptrunner/wptrunner/wpttest.py:557]: It's basically just scafolding for the possibility of implementing this in the future. My guess is that different page sizes will be more important for paginated media than for visual.

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (stephenmcgruer) [reply tools/wptrunner/wptrunner/browsers/chrome.py:86]: Ah, I see. Can you add a comment here noting that (maybe linking crbug.com/753118).

cc @foolip on this, since he's in the webdriver space these days and may know which devtools person to mention this to ;)

Comment (foolip) [reply tools/wptrunner/wptrunner/browsers/chrome.py:87]: I'm afraid is the first time I've seen crbug.com/753118. While I don't know more than what's on the bug, I've added it to a list of maybe good starter bugs for someone joining soonish.

File: tools/wptrunner/wptrunner/wpttest.py
Comment (stephenmcgruer) [reply tools/wptrunner/wptrunner/wpttest.py:558]: Hrm, ok. I prefer not supporting code that isn't used, but I'll leave it up to you to decide.

File: tools/wptrunner/wptrunner/wpttest.py
Comment (stephenmcgruer) [reply tools/wptrunner/wptrunner/wpttest.py:602]: I don't follow here - which underlying API are you referring to?

File: tools/wptrunner/wptrunner/executors/executormarionette.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/executormarionette.py:862]: Nit; should this be once whitespace-line between get_implementation and setup?

File: tools/wptrunner/wptrunner/executors/base.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/base.py:101]: Highlighting a TODO (I'm sure you're aware of it) that I think should be resolved before this is merged.

File: tools/manifest/sourcefile.py
Comment (stephenmcgruer) [inline tools/manifest/sourcefile.py:868]: Oof, nice catch. Could we add a test so future-us don't get hurt when we refactor this?

**Review by stephenmcgruer — COMMENTED** [review]
I think this is nearly there. Though I think I also reviewed this just before the latest force-push, not sure... so if any of my comments are weird I'm going to blame that and not my tiny brain :)

File: tools/wptrunner/wptrunner/executors/base.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/base.py:427]:
```suggestion
        difference occurs if any, or None if there are no differences"""
```

File: infrastructure/reftest/reftest_match-print-ref.html
Comment (stephenmcgruer) [inline infrastructure/reftest/reftest_match-print-ref.html:4]: Note that there's a comma here, is that valid as a CSS separator?

**Review by stephenmcgruer — APPROVED** [review]

File: infrastructure/reftest/reftest_match-print-ref.html
Comment (jgraham) [reply infrastructure/reftest/reftest_match-print-ref.html]: It's not, good catch. But fixing it doesn't seem to change anything.

**Pull request merged**
